File: bot/bot.js
/**
 * Elite India Roleplay - Discord Bot Module
 * 
 * Handles role assignment for passed quiz users
 * 
 * Commands:
 * - !verify - Check your own verification status
 * - !sync - (Admin only) Bulk sync all passed users
 */

const { Client, GatewayIntentBits, PermissionFlagsBits, EmbedBuilder } = require('discord.js');
const storage = require('../src/utils/storage');

// Bot configuration from environment
const BOT_TOKEN = process.env.DISCORD_BOT_TOKEN;
const GUILD_ID = process.env.DISCORD_GUILD_ID;
const ALLOWLIST_ROLE_ID = process.env.ALLOWLIST_ROLE_ID;
const ADMIN_ROLE_ID = process.env.ADMIN_ROLE_ID;
const BOT_PREFIX = process.env.BOT_PREFIX || '!';

// Create Discord client
const client = new Client({
    intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMembers,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
    ],
});

/**
 * Check if a member has admin permissions
 */
function isAdmin(member) {
    if (member.permissions.has(PermissionFlagsBits.Administrator)) {
        return true;
    }
    if (ADMIN_ROLE_ID && member.roles.cache.has(ADMIN_ROLE_ID)) {
        return true;
    }
    return false;
}

/**
 * Assign allowlist role to a member
 */
async function assignRole(member) {
    if (!ALLOWLIST_ROLE_ID) {
        return { success: false, reason: 'Role ID not configured' };
    }

    try {
        const role = member.guild.roles.cache.get(ALLOWLIST_ROLE_ID);
        if (!role) {
            return { success: false, reason: 'Role not found in server' };
        }

        if (member.roles.cache.has(ALLOWLIST_ROLE_ID)) {
            return { success: true, reason: 'Already has role' };
        }

        await member.roles.add(role);
        return { success: true, reason: 'Role assigned' };
    } catch (error) {
        console.error(`[Bot] Failed to assign role to ${member.user.tag}:`, error.message);
        return { success: false, reason: error.message };
    }
}

/**
 * Handle !verify command
 */
async function handleVerify(message) {
    const discordId = message.author.id;
    const user = storage.getUser(discordId);

    const embed = new EmbedBuilder()
        .setColor(0x9b4dca)
        .setTitle('🎮 Elite India Roleplay - Verification')
        .setTimestamp();

    if (!user) {
        embed.setColor(0xff4757)
            .setDescription('❌ You have not registered on the quiz website yet.')
            .addFields({ name: 'Next Steps', value: 'Visit the website and login with Discord to take the quiz.' });
    } else if (user.status === 'passed') {
        embed.setColor(0x00ff88)
            .setDescription('✅ You have passed the allowlist quiz!')
            .addFields({ name: 'Status', value: 'Passed', inline: true });

        // Try to assign role
        const member = message.member;
        if (member && ALLOWLIST_ROLE_ID) {
            const result = await assignRole(member);
            if (result.success) {
                embed.addFields({ name: 'Role', value: result.reason === 'Already has role' ? '✅ Already assigned' : '✅ Assigned!', inline: true });
            } else {
                embed.addFields({ name: 'Role', value: `⚠️ ${result.reason}`, inline: true });
            }
        }
    } else if (user.status === 'failed') {
        const { inCooldown, remainingTime } = storage.isInCooldown(discordId);
        embed.setColor(0xff4757)
            .setDescription('❌ You have not passed the quiz yet.');

        if (inCooldown) {
            const hours = Math.floor(remainingTime / (1000 * 60 * 60));
            const minutes = Math.floor((remainingTime % (1000 * 60 * 60)) / (1000 * 60));
            embed.addFields({ name: 'Cooldown', value: `You can retry in ${hours}h ${minutes}m`, inline: true });
        } else {
            embed.addFields({ name: 'Status', value: 'Cooldown ended. You can retry!', inline: true });
        }
    } else {
        embed.setColor(0xffa502)
            .setDescription('⏳ You have not attempted the quiz yet.')
            .addFields({ name: 'Next Steps', value: 'Visit the website to take the quiz.' });
    }

    await message.reply({ embeds: [embed] });
}

/**
 * Handle !sync command (Admin only)
 */
async function handleSync(message) {
    if (!isAdmin(message.member)) {
        const embed = new EmbedBuilder()
            .setColor(0xff4757)
            .setTitle('❌ Access Denied')
            .setDescription('You do not have permission to use this command.');
        await message.reply({ embeds: [embed] });
        return;
    }

    if (!ALLOWLIST_ROLE_ID) {
        const embed = new EmbedBuilder()
            .setColor(0xff4757)
            .setTitle('❌ Configuration Error')
            .setDescription('ALLOWLIST_ROLE_ID is not configured in .env');
        await message.reply({ embeds: [embed] });
        return;
    }

    const statusMsg = await message.reply('🔄 Syncing roles for all passed users...');

    try {
        const users = storage.getUsers();
        const passedUsers = Object.values(users).filter(u => u.status === 'passed');

        let assigned = 0;
        let alreadyHad = 0;
        let failed = 0;
        let notInServer = 0;

        for (const user of passedUsers) {
            try {
                const member = await message.guild.members.fetch(user.discordId).catch(() => null);

                if (!member) {
                    notInServer++;
                    continue;
                }

                const result = await assignRole(member);
                if (result.success) {
                    if (result.reason === 'Already has role') {
                        alreadyHad++;
                    } else {
                        assigned++;
                    }
                } else {
                    failed++;
                }
            } catch (err) {
                failed++;
            }
        }

        const embed = new EmbedBuilder()
            .setColor(0x00ff88)
            .setTitle('✅ Sync Complete')
            .setDescription(`Processed ${passedUsers.length} passed users`)
            .addFields(
                { name: '✅ Newly Assigned', value: String(assigned), inline: true },
                { name: '✔️ Already Had Role', value: String(alreadyHad), inline: true },
                { name: '❌ Failed', value: String(failed), inline: true },
                { name: '👻 Not in Server', value: String(notInServer), inline: true }
            )
            .setTimestamp();

        await statusMsg.edit({ content: null, embeds: [embed] });

    } catch (error) {
        console.error('[Bot] Sync error:', error);
        await statusMsg.edit('❌ Sync failed: ' + error.message);
    }
}

/**
 * Bot event handlers
 */
client.once('ready', () => {
    console.log(`[Bot] Logged in as ${client.user.tag}`);

    // Set bot activity
    client.user.setActivity('Elite India Roleplay', { type: 3 }); // Watching
});

client.on('messageCreate', async (message) => {
    // Ignore DMs
    if (!message.guild) return;

    // Ignore bots
    if (message.author.bot) return;

    // Check for command prefix
    if (!message.content.startsWith(BOT_PREFIX)) return;

    const args = message.content.slice(BOT_PREFIX.length).trim().split(/\s+/);
    const command = args.shift().toLowerCase();

    try {
        switch (command) {
            case 'verify':
                await handleVerify(message);
                break;
            case 'sync':
                await handleSync(message);
                break;
            case 'help':
                const helpEmbed = new EmbedBuilder()
                    .setColor(0x9b4dca)
                    .setTitle('🎮 Elite India Roleplay Bot')
                    .setDescription('Commands for the allowlist verification system')
                    .addFields(
                        { name: `${BOT_PREFIX}verify`, value: 'Check your quiz status and get your role', inline: false },
                        { name: `${BOT_PREFIX}sync`, value: '(Admin) Bulk assign roles to all passed users', inline: false }
                    )
                    .setFooter({ text: 'Complete the quiz on the website to get allowlisted!' });
                await message.reply({ embeds: [helpEmbed] });
                break;
        }
    } catch (error) {
        console.error(`[Bot] Error handling command ${command}:`, error);
    }
});

client.on('error', (error) => {
    console.error('[Bot] Discord client error:', error);
});

/**
 * Start the Discord bot
 */
async function startBot() {
    if (!BOT_TOKEN) {
        throw new Error('DISCORD_BOT_TOKEN is not set in .env');
    }

    await client.login(BOT_TOKEN);

    // Wait for ready event
    return new Promise((resolve) => {
        if (client.isReady()) {
            resolve(client);
        } else {
            client.once('ready', () => resolve(client));
        }
    });
}

/**
 * Get the bot client instance
 */
function getClient() {
    return client;
}

/**
 * Assign role to a user by Discord ID (called from website when quiz passed)
 * @param {string} discordId - The Discord user ID
 * @returns {Promise<{success: boolean, reason: string}>}
 */
async function assignRoleToUser(discordId) {
    if (!client.isReady()) {
        console.log('[Bot] Cannot assign role - bot not ready');
        return { success: false, reason: 'Bot not ready' };
    }

    if (!GUILD_ID) {
        console.log('[Bot] Cannot assign role - DISCORD_GUILD_ID not configured');
        return { success: false, reason: 'Guild ID not configured' };
    }

    if (!ALLOWLIST_ROLE_ID) {
        console.log('[Bot] Cannot assign role - ALLOWLIST_ROLE_ID not configured');
        return { success: false, reason: 'Role ID not configured' };
    }

    try {
        // Get the guild
        const guild = await client.guilds.fetch(GUILD_ID);
        if (!guild) {
            return { success: false, reason: 'Guild not found' };
        }

        // Get the member
        const member = await guild.members.fetch(discordId).catch(() => null);
        if (!member) {
            return { success: false, reason: 'User not in server' };
        }

        // Get the role
        const role = guild.roles.cache.get(ALLOWLIST_ROLE_ID);
        if (!role) {
            return { success: false, reason: 'Role not found' };
        }

        // Check if already has role
        if (member.roles.cache.has(ALLOWLIST_ROLE_ID)) {
            console.log(`[Bot] User ${member.user.tag} already has allowlist role`);
            return { success: true, reason: 'Already has role' };
        }

        // Assign the role
        await member.roles.add(role);
        console.log(`[Bot] ✅ Assigned allowlist role to ${member.user.tag}`);

        // Optional: Send DM to user
        try {
            await member.send({
                embeds: [{
                    color: 0x00ff88,
                    title: '🎉 Congratulations!',
                    description: 'You have passed the Elite India Roleplay allowlist quiz!',
                    fields: [
                        { name: 'Status', value: '✅ Allowlisted', inline: true },
                        { name: 'Role', value: '✅ Assigned', inline: true }
                    ],
                    footer: { text: 'Welcome to Elite India Roleplay!' },
                    timestamp: new Date().toISOString()
                }]
            });
        } catch (dmError) {
            // User has DMs disabled, that's okay
            console.log(`[Bot] Could not DM user ${member.user.tag} (DMs disabled)`);
        }

        return { success: true, reason: 'Role assigned' };

    } catch (error) {
        console.error(`[Bot] Error assigning role to ${discordId}:`, error.message);
        return { success: false, reason: error.message };
    }
}

/**
 * Remove allowlist role from a user by Discord ID (called when admin fails user)
 * @param {string} discordId - The Discord user ID
 * @returns {Promise<{success: boolean, reason: string}>}
 */
async function removeRoleFromUser(discordId) {
    if (!client.isReady()) {
        console.log('[Bot] Cannot remove role - bot not ready');
        return { success: false, reason: 'Bot not ready' };
    }

    if (!GUILD_ID) {
        console.log('[Bot] Cannot remove role - DISCORD_GUILD_ID not configured');
        return { success: false, reason: 'Guild ID not configured' };
    }

    if (!ALLOWLIST_ROLE_ID) {
        console.log('[Bot] Cannot remove role - ALLOWLIST_ROLE_ID not configured');
        return { success: false, reason: 'Role ID not configured' };
    }

    try {
        // Get the guild
        const guild = await client.guilds.fetch(GUILD_ID);
        if (!guild) {
            return { success: false, reason: 'Guild not found' };
        }

        // Get the member
        const member = await guild.members.fetch(discordId).catch(() => null);
        if (!member) {
            return { success: false, reason: 'User not in server' };
        }

        // Get the role
        const role = guild.roles.cache.get(ALLOWLIST_ROLE_ID);
        if (!role) {
            return { success: false, reason: 'Role not found' };
        }

        // Check if user has the role
        if (!member.roles.cache.has(ALLOWLIST_ROLE_ID)) {
            console.log(`[Bot] User ${member.user.tag} doesn't have allowlist role`);
            return { success: true, reason: 'Role not present' };
        }

        // Remove the role
        await member.roles.remove(role);
        console.log(`[Bot] ❌ Removed allowlist role from ${member.user.tag}`);

        // Optional: Send DM to user
        try {
            await member.send({
                embeds: [{
                    color: 0xff4757,
                    title: '⚠️ Allowlist Status Changed',
                    description: 'Your allowlist status has been revoked by an administrator.',
                    fields: [
                        { name: 'Status', value: '❌ Revoked', inline: true },
                        { name: 'Action Required', value: 'Please retake the quiz', inline: true }
                    ],
                    footer: { text: 'Elite India Roleplay' },
                    timestamp: new Date().toISOString()
                }]
            });
        } catch (dmError) {
            console.log(`[Bot] Could not DM user ${member.user.tag} (DMs disabled)`);
        }

        return { success: true, reason: 'Role removed' };

    } catch (error) {
        console.error(`[Bot] Error removing role from ${discordId}:`, error.message);
        return { success: false, reason: error.message };
    }
}

/**
 * Notify user of staff application result
 */
async function notifyStaffAppResult(discordId, status, reason) {
    if (!client.isReady()) return { success: false, reason: 'Bot not ready' };

    try {
        const user = await client.users.fetch(discordId).catch(() => null);
        if (!user) return { success: false, reason: 'User not found' };

        const isAccepted = status === 'accepted';
        const embed = new EmbedBuilder()
            .setColor(isAccepted ? 0x00ff88 : 0xff4757)
            .setTitle(isAccepted ? '👔 Staff Application Accepted!' : '👔 Staff Application Rejected')
            .setDescription(`Your staff application for Elite India Roleplay has been ${status}.`)
            .addFields({ name: 'Reason/Message', value: reason || 'No specific reason provided.' })
            .setTimestamp();

        if (isAccepted) {
            embed.addFields({ name: 'Next Steps', value: 'Please contact a Senior Admin for your onboarding.' });
        }

        await user.send({ embeds: [embed] });
        return { success: true };
    } catch (error) {
        console.error(`[Bot] Failed to DM user ${discordId}:`, error.message);
        return { success: false, reason: error.message };
    }
}

/**
 * Notify user of gang application result
 */
async function notifyGangAppResult(discordId, status, gangName, reason) {
    if (!client.isReady()) return { success: false, reason: 'Bot not ready' };

    try {
        const user = await client.users.fetch(discordId).catch(() => null);
        if (!user) return { success: false, reason: 'User not found' };

        const isAccepted = status === 'accepted';
        const embed = new EmbedBuilder()
            .setColor(isAccepted ? 0x00ff88 : 0xff4757)
            .setTitle(isAccepted ? '🔫 Gang Application Accepted!' : '🔫 Gang Application Rejected')
            .setDescription(`Your application for gang **${gangName}** has been ${status}.`)
            .addFields({ name: 'Reason/Message', value: reason || 'No specific reason provided.' })
            .setTimestamp();

        if (isAccepted) {
            embed.addFields({ name: 'Next Steps', value: 'Your gang registration is now official. Please check gang rules for more info.' });
        }

        await user.send({ embeds: [embed] });
        return { success: true };
    } catch (error) {
        console.error(`[Bot] Failed to DM user ${discordId}:`, error.message);
        return { success: false, reason: error.message };
    }
}

module.exports = { startBot, getClient, assignRoleToUser, removeRoleFromUser, notifyStaffAppResult, notifyGangAppResult };

File: bot/bot.py
import os
import asyncio
import discord
from discord.ext import commands
from supabase import create_client, Client
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configuration
BOT_TOKEN = os.getenv('DISCORD_BOT_TOKEN')
SUPABASE_URL = os.getenv('SUPABASE_URL')
SUPABASE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
GUILD_ID = int(os.getenv('DISCORD_GUILD_ID', 0))
ALLOWLIST_ROLE_ID = int(os.getenv('ALLOWLIST_ROLE_ID', 0))
ADMIN_ROLE_ID = int(os.getenv('ADMIN_ROLE_ID', 0))
PREFIX = os.getenv('BOT_PREFIX', '!')

# Initialize Supabase
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

# Initialize Bot
intents = discord.Intents.default()
intents.members = True
intents.message_content = True
bot = commands.Bot(command_prefix=PREFIX, intents=intents)

async def assign_role_to_member(member: discord.Member):
    """Assigns the allowlist role to a member."""
    if not ALLOWLIST_ROLE_ID:
        return False, "Role ID not configured"
    
    role = member.guild.get_role(ALLOWLIST_ROLE_ID)
    if not role:
        return False, "Role not found in server"
    
    if role in member.roles:
        return True, "Already has role"
    
    try:
        await member.add_roles(role)
        return True, "Role assigned"
    except Exception as e:
        print(f"[Bot] Error assigning role to {member}: {e}")
        return False, str(e)

@bot.event
async def on_ready():
    print(f'[Bot] Logged in as {bot.user.name} ({bot.user.id})')
    await bot.change_presence(activity=discord.Activity(type=discord.ActivityType.watching, name="Elite India Roleplay"))
    
    # Start Realtime Listener in background
    bot.loop.create_task(listen_for_passed_quiz())

async def listen_for_passed_quiz():
    """Listens for new 'passed' attempts in Supabase Realtime."""
    print("[Bot] Starting Realtime listener for 'attempts' table...")
    
    def on_handle(payload):
        # This is the callback for realtime events
        new_record = payload.get('record')
        if not new_record:
            return
            
        is_passed = new_record.get('passed')
        discord_id = new_record.get('discord_id')
        
        if is_passed and discord_id:
            print(f"[Bot] Realtime: User {discord_id} passed!")
            # We need to run the role assignment in the bot's event loop
            asyncio.run_coroutine_threadsafe(auto_assign_role(discord_id), bot.loop)

    try:
        # Note: supabase-py Realtime support can be tricky depending on version
        # Recommended: Use the storage utility patterns if realtime is not enabled or stable
        # For now, we subscribe to 'attempts' INSERTs
        supabase.table('attempts').on_insert(on_handle).subscribe()
    except Exception as e:
        print(f"[Bot] Realtime Error: {e}")

async def auto_assign_role(discord_id):
    """Automatically assigns role to a user given their Discord ID."""
    guild = bot.get_guild(GUILD_ID)
    if not guild:
        print(f"[Bot] Guild {GUILD_ID} not found")
        return

    member = guild.get_member(int(discord_id))
    if not member:
        # Try to fetch if not in cache
        try:
            member = await guild.fetch_member(int(discord_id))
        except:
            print(f"[Bot] User {discord_id} not in server")
            return

    success, reason = await assign_role_to_member(member)
    if success:
        print(f"[Bot] Auto-role: Assigned to {member.name} ({reason})")
        # Notify user in DM
        try:
            embed = discord.Embed(
                title="🎉 Congratulations!",
                description="You have passed the Elite India Roleplay allowlist quiz!",
                color=0x00ff88
            )
            embed.add_field(name="Status", value="✅ Allowlisted")
            embed.set_footer(text="Welcome to Elite India Roleplay!")
            await member.send(embed=embed)
        except:
            pass

@bot.command()
async def verify(ctx):
    """Check your own verification status."""
    discord_id = str(ctx.author.id)
    
    # Query Supabase for user status
    res = supabase.table('users').select('*').eq('discord_id', discord_id).execute()
    user_data = res.data[0] if res.data else None
    
    embed = discord.Embed(title="🎮 Elite India Roleplay - Verification", color=0x9b4dca)
    
    if not user_data:
        embed.description = "❌ You have not registered on the quiz website yet."
        embed.color = 0xff4757
        embed.add_field(name="Next Steps", value="Visit the website and login with Discord to take the quiz.")
    elif user_data.get('status') == 'passed':
        embed.description = "✅ You have passed the allowlist quiz!"
        embed.color = 0x00ff88
        
        success, reason = await assign_role_to_member(ctx.author)
        role_status = "✅ Assigned!" if reason == "Role assigned" else "✅ Already assigned" if success else f"⚠️ {reason}"
        embed.add_field(name="Role Status", value=role_status)
    else:
        embed.description = "⏳ You have not passed the quiz yet."
        embed.add_field(name="Next Steps", value="Visit the website to take/retry the quiz.")

    await ctx.reply(embed=embed)

@bot.command()
@commands.has_permissions(administrator=True)
async def sync(ctx):
    """(Admin only) Bulk sync all passed users."""
    status_msg = await ctx.send("🔄 Syncing roles for all passed users...")
    
    # Fetch all passed users from Supabase
    res = supabase.table('users').select('discord_id').eq('status', 'passed').execute()
    passed_users = res.data
    
    assigned = 0
    already_had = 0
    failed = 0
    not_in_server = 0
    
    for user in passed_users:
        d_id = int(user['discord_id'])
        member = ctx.guild.get_member(d_id)
        if not member:
            try:
                member = await ctx.guild.fetch_member(d_id)
            except:
                not_in_server += 1
                continue
        
        success, reason = await assign_role_to_member(member)
        if success:
            if reason == "Already has role":
                already_had += 1
            else:
                assigned += 1
        else:
            failed += 1
            
    embed = discord.Embed(title="✅ Sync Complete", color=0x00ff88)
    embed.add_field(name="✅ Newly Assigned", value=str(assigned))
    embed.add_field(name="✔️ Already Had Role", value=str(already_had))
    embed.add_field(name="❌ Failed", value=str(failed))
    embed.add_field(name="👻 Not in Server", value=str(not_in_server))
    
    await status_msg.edit(content=None, embed=embed)

if __name__ == '__main__':
    if not BOT_TOKEN:
        print("❌ Error: DISCORD_BOT_TOKEN not found in environment")
    else:
        bot.run(BOT_TOKEN)
